### ISSUE TITLE
Unify after() and timeout() 

### ISSUE THREAD

**Issue by ChristianSchwarz** [mockito/mockito#472]
Mockito 2.0 is coming soon™ and this is a chance to change the API. I would like to see a unification of the two very similar concurrent VerificationModes `After `and `Timeout `into one single VerificationMode.
**Why?**
 * Mockito user's don't know immediatly which VerificationMode to use when, they must read the doc's  carefully.
 * The specialized VerificationModes like never(),onlyOnce() and so on are very similar, the only difference is the fail fast / succeed fast behaviour.
 * There is no reason to wait until the specified time is elapsed if the verifcation can fail or succeed fast.

**How should it look like?**
`verify(mock, within(120,MILLISECONDS)).methodCall();` 
succeed fast if the specified call was detected or fails if the 120ms elapsed and no call was detected

`verify(mock, within(120,MILLISECONDS).atMost(5)).methodCall();`
fails fast if within 120ms more than 5 calls were detected, succeed after 120ms if at most 5 calls were detected

**How to realize it mockito 2.0 is scheduled for 31July?**
The Within-VerificationMode can reuse/deleagte large parts of/to the existing Timeout and After implementation. This way public API of After/Timeout can be removed and the Within-API can be added with little risk. In the mid term the underlying implementation can be refactored and After/Timeout can possibly removed completly.

**How should the final implementation look like?**
Currently the concurrent verifications are done in tight loops that doesn't only consume CPU time but also have [side effects on argument captors](#379). The within implementation should us a event based verifcation. Simplified said only when a new invocation is detected the VerificationMode should verify it. This way there is no need for workarounds like [removing previously verified invocations](#349).

Any comment is welcome!

**Comment by bric3** [comment]
I think we should postpone this task to 2.1. Thoughts ?

**Comment by TimvdLippe** [comment]
Yes agreed we should not rush this feature for Mockito 2.0

**Comment by szczepiq** [comment]
This is really good feature request, the use cases are laid out very clearly.

Questions bout proposed plan:

1. Replacing both "timeout" _and_ "after" with "await" + fluent atMost/atLeast/etc? +1 to it.
2. Is there a default for TimeUnits? +1 to it.

**Comment by bric3** [comment]
Before going full force on this please check the feedback of @jmborer here https://github.com/mockito/mockito/issues/345#issuecomment-241032338

**Comment by ChristianSchwarz** [comment]
Here is the part of the comment regarding the after(..) usecase :

> Then I have situations where I want to wait a given time and THEN only do the check. This is where I use after() instead of Thread.sleep()+verify() which I find less fluent to read...

If a user don't want to succeed or fail fast then sleep()+verify() should be choosen. I think it is very expressive and self explaining. All other cases should be covered by await().

**Comment by jmborer** [comment]
I am sorry, but I find "await" very vague and ambiguous (therefore error prone). Await, ok but what? Await until a criteria is met, or await until a given timeout is reached?

timeout and after are far more dis-ambiguous IMHO. With "await" I will have to check the doc to be sure to understand what it does. For timeout and after, I read it only once...

My two cents (but honestly I won't be happy if those methods disappear)

**Comment by ChristianSchwarz** [comment]
@jmborer 
Thanks for you feedback! 

> I am sorry, but I find "await" very vague and ambiguous (therefore error prone). Await, ok but what? Await until a criteria is met, or await until a given timeout is reached?

Other proposals are "within" or "during" (see #91). I choose "await" here cause it is a well known concept from the concurrency APIs (Condition.await(..), Future.await(..), Lock.await(..)) with the same behavior. A criteria was met or can't be satisfied and the method returns/throws immediatly or a timeout was reached.

Why do you need two different methods "timeout"/"after"? If I want to check if call were made after some time elapsed I rather would write something like this:
```
 sleep(50,MILLIS);
 verify(mock).call(123);
```

**Comment by jmborer** [comment]
Ok so "await" is equivalent to the former "timeout": it is blocked until the verification is satisfied or timeout is reached. In other words, as soon as the verification is satisfied, it is unblocked or exits at timeout. Right?

Concerning "after", I agree it is equivalent to sleep+verify. Do I have to catch interruption exception for the "sleep"? If no, then it could be ok. Maybe it is actually easier to read "sleep+verify" instead of understanding the difference between _timeout_ and _after_ then.

**Comment by ChristianSchwarz** [comment]
> Ok so await is equivalent to the former timeout: it is blocked until the verification is satisfied or timeout is reached. In other words, as soon as the verification is satisfied, it is unblocked or exits at timeout. Right?

Yes "await" returns in three different cases:
 * the verification is satisfied within time
 * the verification failed within time
 * the verification failed caused by a timeout

The difference is that "await" will also supports _atMost(max)_. 

---

> Concerning after, I agree it is equivalent to sleep+verify. Do I have to catch interruption exception for the sleep call (as I had to do with Thread.sleep)? 

You can add InterruptedException to your test method declaration or uses guavas [sleepUninterruptibly](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Uninterruptibles.html#sleepUninterruptibly(long,%20java.util.concurrent.TimeUnit)).

---

`Maybe it is actually easier to read sleep+verify instead of understanding the difference between timeout and after.`

Yes thats the point! A simple API = better user experience.


> What about another option, more fluent that would look like:
> 
> after(50,MILLIS).verify(mock).call(123)

It looks neat but the devil is in the details. E.g how would you write "inorder.verifiy(mock,after(100).atLeast(5)).call(123)" this? It is possible but this is a huge amout of refactoring.

**Comment by szczepiq** [comment]
Great discussion!

>after(50,MILLIS).verify(mock).call(123)

Please no.

This is adding new public API for something that can be easily implemented already (sleep() followed by verify()). Such construct would be used rarely. Users can always create a handy method like verifyAfter() and reduce duplication when needed.

**Comment by jmborer** [comment]
> after(50,MILLIS).verify(mock).call(123)

Don't be afraid. It was more of a brain storm.

Anyway, wouldn't it be possible to deprecated _timeout_ and _after_ in Mockito 2.0, but still leave them so that we can migrate the tests and remove them definitely in version 2.1? I use them a lot and search/replace won't help me easily here.

**Comment by bric3** [comment]
@jmborer Nothing is set in stone yet on this. But be sure we won't retire these API without deprecating them beforehand and we certainly deprecated them without an alternative.

> Await, ok but what? Await until a criteria is met, or await until a given timeout is reached?

@ChristianSchwarz proposed the following API, could it be possible to include remarks of @jmborer 

```java
verify(mock, await(120,MILLISECONDS).atMost(5)).methodCall();
```

I certainly don't like to declare `InterruptedException` in my tests, like it's a normal thing for a test to declare it _throws_ exception(s). I'd rather like somehting included in the mockito API.

```java
verify(mock, after(10, SECONDS).await(120,MILLISECONDS).atMost(5)).methodCall();
```

**Comment by ChristianSchwarz** [comment]
Hi @bric3,
 
`verify(mock, after(10, SECONDS).await(120,MILLISECONDS).atMost(5)).methodCall();` doesn't read well and I am afraid that it doesn't reveal its intend. 

In case someone need's a thread to sleep without the need to declare or catch an `InterruptedException` i would propose to use guava's [sleepUninterruptibly(..)](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Uninterruptibles.html#sleepUninterruptibly(long,%20java.util.concurrent.TimeUnit)) or to implement similar functionality itself.
